### PR TITLE
fix/FD-3355-redirect-to-path-after-2fa

### PR DESCRIPTION
**Pull request by far-fetched**

### CODE REVIEW

**Comment by Bojakowski** [comment]
Hmm, I think it looks good, however I'm wondering if `redirect_path` could somehow redirect you to place where you don't want to go. What does this variable store, and where the value comes from?

**Comment by werleo** [comment]
@dominika will you share your opinion please.

File: lib/devise_google_authenticatable/patches/check_ga.rb [hunk 1]
```diff
@@ -17,10 +21,11 @@ module CheckGA
 
         resource = warden.authenticate!(:scope => resource_name, :recall => "#{controller_path}#new")
 
+        redirect_path = after_sign_in_path_for(resource)
         if resource.respond_to?(:get_qr) and resource.gauth_enabled? and resource.require_token?(cookies.signed[:gauth]) #Therefore we can quiz for a QR
           tmpid = resource.assign_tmp #assign a temporary key and fetch it
           warden.logout #log the user out
-
+          cache_in_session_redirect_path(redirect_path)
```
Comment (dominika) [inline lib/devise_google_authenticatable/patches/check_ga.rb]: do you think we could use the devise's [store_location_for](https://www.rubydoc.info/github/plataformatec/devise/Devise%2FControllers%2FStoreLocation:store_location_for) – which puts the path in the session store –we would not need to define our own cache method for this and probably would not need the additional condition in the checkga controller `session[:cached_redirect_path] ||`

did you maybe tried it?

File: app/controllers/devise/checkga_controller.rb [hunk 2]
```diff
@@ -21,7 +21,7 @@ def update
         set_flash_message(:notice, :signed_in) if is_navigational_format?
         sign_in(resource_name,resource)
         warden.manager._run_callbacks(:after_set_user, resource, warden, {:event => :authentication})
-        respond_with resource, :location => after_sign_in_path_for(resource)
+        respond_with resource, :location => stored_location_for(resource) || after_sign_in_path_for(resource)
```
Comment (dominika) [inline app/controllers/devise/checkga_controller.rb]: is it necessary to use the `stored_location_for`? I do not remember the implementation of the `after_sign_in_path` but maybe it used the stored path anyway? any chance it does :D?

Comment (werleo) [reply app/controllers/devise/checkga_controller.rb]: @dominika you are right about it

https://www.rubydoc.info/github/plataformatec/devise/Devise/Controllers/Helpers:after_sign_in_path_for

Comment (dominika) [reply app/controllers/devise/checkga_controller.rb]: that's what I thought! 
https://img.memecdn.com/delete-all-the-things_o_4994465.jpg

**Comment by dominika** [comment]
hmmm? are those any tests for this?

**Comment by burczan** [comment]
> hmmm? are there any tests for this?

Please let me know if you prefer to write acceptance/integration tests for that :)

**Comment by dominika** [comment]
a couple of tests from the author @far-fetched I suppose right here :)

**Pull request merged**